app/uk/gov/hmrc/transitmovements/controllers/ObjectStoreURIHeaderExtractor.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovements.controllers
+
+import cats.data.EitherT
+import play.api.mvc.Headers
+import uk.gov.hmrc.transitmovements.config.Constants
+import uk.gov.hmrc.transitmovements.controllers.errors.HeaderExtractError
+import uk.gov.hmrc.transitmovements.controllers.errors.HeaderExtractError.InvalidObjectStoreURI
+import uk.gov.hmrc.transitmovements.controllers.errors.HeaderExtractError.NoHeaderFound
+import uk.gov.hmrc.transitmovements.models.ObjectStoreURI
+
+import scala.concurrent.Future
+
+trait ObjectStoreURIHeaderExtractor {
+
+  def extractObjectStoreURI(headers: Headers): EitherT[Future, HeaderExtractError, ObjectStoreURI] =
+    EitherT {
+      headers.get(Constants.ObjectStoreURI) match {
+        case None => Future.successful(Left(NoHeaderFound("Missing X-Object-Store-Uri header value")))
+        case Some(headerValue) =>
+          ObjectStoreURI(headerValue).path match {
+            case None                 => Future.successful(Left(InvalidObjectStoreURI(s"Invalid X-Object-Store-Uri header value: $headerValue")))
+            case Some(objectStoreURI) => Future.successful(Right(ObjectStoreURI(objectStoreURI)))
+          }
+      }
```

Comment:
You've created another problem here by doing this now, `ObjectStoreURI` now represents both the full URI, and the fragment that does not contain `common-transit-convention-traders`, and that's bad because `ObjectStoreURI` was created as a marker to indicate a specific type of value... now you have two types of values under the same marker.

I also suggested what I did before because you were passing this whole URI to a service, which could then get the information from the value. You're now using `ObjectStoreURI` to store the computed value, so `path` shouldn't exist on it.

If you're going to go down this route, remove `path` from `ObjectStoreURI` and do it in here instead. Something like the following will possibly work:

```scala
object ObjectStoreURIHeaderExtractor {

  // The URI consists of the service name in the first part of the path, followed
  // by the location of the object in the context of that service. As this service
  // targets common-transit-convention-traders' objects exclusively, we ensure
  // the URI is targeting that context. This regex ensures that this is the case.
  val expectedUriPattern: Regex = "^common-transit-convention-traders/(.+)$".r
}
trait ObjectStoreURIHeaderExtractor {

  def extractObjectStoreURI(headers: Headers): EitherT[Future, PresentationError, ObjectStoreURI] =
    EitherT(
      Future.successful(
        for {
          headerValue    <- headers.get(Constants.ObjectStoreURI).toRight(PresentationError.badRequestError("Missing X-Object-Store-Uri header value"))
          objectStoreUri <- expectedUriPattern.findFirstMatchIn(headerValue).map(_.group(1)).toRight(PresentationError.badRequestError(s"X-Object-Store-Uri header value does not start with common-transit-convention-traders/ (got $headerValue)"))
        } yield ObjectStoreURI(objectStoreUri)
      )
    )
}
```

Note the following:

* `ObjectStoreURI` is a bad name now, I suggest something like `ObjectStoreResourceLocation` (because this object will no longer hold a full URI).
* This trait is intended to be mixed into a presentation layer object, because you're extracting from a header. In this case, just use `PresentationError` objects directly. Liandra self typed her presentation layer trait to `BaseController`, you could do that too. 
* You can split the two lines out into their own functions if you want.
* Check what I've written is correct and don't take it for granted, it's not completely tested.